### ISSUE TITLE
Setting user per-host isn't working

### ISSUE THREAD

**Issue by maxim** [capistrano/capistrano#1427]

**Comment by leehambley** [comment]
It likely has to do with the changes to server resolution. One workaround would be to specify the username and domain name in a global (`~/.ssh/config`) or a project local `~/yourproh/.ssh/config`. See https://github.com/capistrano/sshkit/blob/5790401bf5a7f9bdbf2d9ff6f5867001ce465c66/CHANGELOG.md#120, I thought it was documented somewhere else, now I see it's only in the changelog, no wonder nobody knows about it :)

It's a shitty workaround, I know, but this way the hosts will look totally different to sshkit:

```
Host 1
   User userone
   Host hostone

Host 2
   User usertwo
   Host hostone
```

Then connect to `1` and `2` (choose better names!) in SSHKit.

**Comment by leehambley** [comment]
You might also try:

``` ruby
server 'deployer@foo.example.org', roles: [:full]
```

**Comment by maxim** [comment]
@leehambley thanks for alternatives. Your latter `'deployer@…'` version didn't help, but I realized that if I comment out the 2nd declaration in my original example like this:

```ruby
server 'foo.example.org', user: 'deployer', roles: [:util]
# role :util_exec, 'foo.example.org', user: 'runner', no_release: true
```

then `whoami` actually shows deployer as expected. So having 1 server in 2 roles with different user assigned to it is no longer possible this way (e.g. for deployment tasks you need deployer user but for running console, scripts, etc you need runner user on the same host). You're saying .ssh/config is the workaround? I agree it's far from ideal, at least because you now gotta specify your servers in multiple places. Any plans to restore the old behavior in this case?

**Comment by leehambley** [comment]
The server matching code has undergone quite some changes, you can check the code to see who the "Blame" points to and bring the discussion back here. If you can write a test case for your specific problem using our suite (there's plenty of reference code) then we have a concrete place to start. 

It seems however that whatever we do with "merge the servers intelligently" vs. "last value wins" vs. "allow multiple" we'll always get it wrong for some people.

I don't know that I agree that the ssh config is a bad workaround, I think the following is kinda neat:

``` 
# ssh config in ./myproject/.ssh/config
Host util_server
    Host foo.example.org
    User runner
```

paired with a single line change:

``` diff
- role :util_exec, 'foo.example.org', user: 'runner', no_release: true
+ role :util_exec, util_server, no_release: true
```

is all that bad?

**Comment by leehambley** [comment]
I just looked it up, the server "identity" matching code is all here - https://github.com/capistrano/capistrano/blob/master/spec/lib/capistrano/configuration/server_spec.rb#L35

**Comment by maxim** [comment]
> I don't know that I agree that the ssh config is a bad workaround

Well, you did say (and I quote) "it's a shitty workaround". :)  But yeah maybe it is kind of neat to have it set like that. Problem is — isn't that what roles are for? Adding `Host util_server` is like saying "in my stage this is what plays the role of util server". That's directly overlapping host/role declarations in capistrano. Given that, what should this code do then? Why even have this code?

```ruby
server 'foo.example.org', user: 'deployer', roles: [:util]
server 'foo.example.org', user: 'runner',   roles: [:exec]
```

**Comment by leehambley** [comment]
Haha, I guess my opinion swung, after I wrote your comment I added [this section](https://github.com/capistrano/sshkit/blob/master/README.md#tunneling-and-other-related-ssh-themes) to the README which was sorely missing.

This is a good case, too, and a case where I think if you were to ask a **n** programmers their opinion, you'd get **n+1** answers.

A more interesting case, is where do you draw the line:

``` ruby
server 'foo.example.org', user: 'deployer', roles: [:util], port: 123
server 'foo.example.org', user: 'runner',   roles: [:exec], port: 456
```

Maybe… one could argue if the ports differ, the "host" differs, and we normalize based on "host", but, then maybe to do what you want the only valid syntax might be:

``` ruby
server 'deployer@foo.example.org:123', user: 'deployer', roles: [:util]
server 'runner@foo.example.org:456', user: 'runner',   roles: [:exec]
```

The server "equality" logic for the de-duplification is kind of tricky, as I mentioned before I'm not sure anyone could really agree on how we'd expect it to work :) I don't even remember how it was originally designed, I *believe* I wrote it to take the *connection* properties into account, i.e user, hostname, port, but it's been changed since as that became problematic when people wanted to mix the `server …` syantax with the `role …` syntax, using the latter to add arbitrary properties to the servers defined in that role.

I guess in the end that's the value to providing a multitude of APIs, but the number of complaints has gone up about the server matching semantics *since* the change, now I can't say with any degree of certainty whether that's because it's *changed* or whether it's because it's **worse**.

**Comment by leehambley** [comment]
In lieu of other test cases, please, feel free to submit your "wishlist" of server matching cases, if the behaviour is poorly or presently undefined, you have a chance to plant your flag, and since we try very hard not to *break* APIs for which we have *tests* you might get what you want!

**Comment by maxim** [comment]
I would do that, but I'm not sure myself what exactly is the right way. What is host identity? What if you have multiple VMs running on different ports of the same hostname? Are those the same host? I don't think so. I believe full host identity is what you wrote above: [user, hostname, port], e.g.:

```ruby
server 'deployer@foo.example.org:123'
host.user # => "deployer"
host.name # => "foo.example.org"
host.port # => 123

server 'foo.example.org'
host.user # => "[whatever is default, or setup in ssh]"
host.name # => "foo.example.org"
host.port # => [default ssh port]
```

So even if port/user are omitted, the host still has them, just defaults. So if you want to add more properties on the same host, you gotta ensure that all 3 things match. This means we can also no longer use separate `host:` and `port:` keys as server attributes, instead only the complete string will serve as host identifier. I'm not saying this is exactly how I'd want it, just thinking out loud. This is hard for such a legacy project, I respect the pains involved in figuring this out.

**Comment by leehambley** [comment]
> What is host identity?

Exactly :)

**Comment by leehambley** [comment]
As I said, the best I can offer you is that if you can create the test case you expect to work, we can work to that implementation (actually, testing it is harder, but anything short of red/green is just a silly way to talk about it) Personally I've never needed any of these things, having always had relatively simple deployments for clients :)

**Comment by dreyks** [comment]
@leehambley 
I've posted a failing test in #784 
Don't know if you get notifications to closed issues. If you did get that too, sorry for the doubleposting

**Comment by leehambley** [comment]
@dreyks please try and be clear about what you want to achieve. Your proposed change broke a selection of unit tests, we can't merge your proposal as long as it breaks our test suite, even if you are proposing another test which would pass under your scheme.

**Comment by dreyks** [comment]
Well, I didn't propose any changes. I simply created a 'case I expect to work'. 

I understand that this is an opinionated topic. There's a handful of issues related to 'single host for different servers' but not everybody needs this feature.

What I'm asking is Are you interested in changing the way the uniqueness of hosts work? If not - we can close this issues and recommend everybody to use project-specific ssh config files. But if you are interested - then we can work together towards resolving this

**Comment by leehambley** [comment]
@dreyks If you propose a non-breaking change, or prepare a pull request that meets our contribution guidelines I'd be more than happy to engage with you to find a solution.

**Comment by tanelsuurhans** [comment]
Bumping this one a little. So the behavior of setting a user per server is imho not desirable in most cases, and people either revert to using agent forwarding or setting a user via ssh options. In both of the latter cases the host user is actually reported as nil (and the configured ssh user is tacked on at a later stage when SSH is invoked). Seems like ideally the host.user should always report the user that will be used when the SSH connection is opened, making it easier to roll out plugins for example.

Right now you need to know that the user has to be checked either from the host or the host ssh options, or in some cases even the backends ssh options.

**Comment by mattbrictson** [comment]
> Seems like ideally the host.user should always report the user that will be used when the SSH connection is opened, making it easier to roll out plugins for example.

Could you elaborate on this use case a bit more?

Unfortunately we can't really know what the user is until the SSH connection is already open, so it is difficult to make this change. But if there is a compelling use case for it I may dig into it some more.

**Comment by laszlof** [comment]
Are there any plans on fixing this? I just ran into this myself and had a heck of a time finding out information on it.

My use case is, I have a security requirement to keep services separated by users. That being said, I need to do things as separate users on the same server. It stands to reason that the following should work:

```
role :app, 'user1@host1.com'
role :something, 'user2@host1.com'

However, this results in everything on both :app and :something role's being executed by user2. The ssh config workaround will not work in my case, as the hosts are varied based on environment and I need this to be easily reproducable without a lot of additional steps and setup reqirements.

**Comment by mattbrictson** [comment]
Capistrano supports loading SSH config from the local project. In other words, within your project, you can place a `./.ssh/config` file and Capistrano will automatically use it when you deploy. That way you can vary the configuration per project, and you can even place it in source control so that other members of your team don't need to do any setup of their own.

Will that work for your case?

**Comment by mattbrictson** [comment]
Just to reiterate/summarize this issue: Capistrano's DSL does not allow declaring two different servers with identical hostnames but different usernames.

There is currently a workaround, using a local `.ssh/config` file, which can be committed as part of the project. This solution is probably "good enough" for most people, because the `.ssh/config` only needs to be set up once when the servers and Capistrano configuration are initially defined. After that, since the config file travels with the repo, it "just works".

However for projects where servers are often being added or changed, or for developers who are often setting up new projects, the `.ssh/config` file is an extra step in the process. It also splits configuration across multiple files, which is not ideal.

I am aware of this, but it is not a high priority for me to fix at the moment. If anyone following this discussion could contribute a PR that adds distinct `user@host` support without breaking existing tests, that would be greatly appreciated!

**Comment by laszlof** [comment]
Right, sorry for the delayed response. Unfortunately the SSH config wont work in my situation. We have multiple stages with different servers, and the servers may change regularly or via the environment.

I'll take a look at the code personally when I have some free time and see if I can work together a PR to fix the issue.

If you have any information that may be useful in finding the place in the code where the modifications need to be done, it would make my job easier as I wouldnt have to fully understand everything before beginning a PR. :)

Thanks again.

**Comment by laszlof** [comment]
I actually managed to fix it fairly easily, adjusted a couple of the tests to fit. But ran into an issue with overriding an existing user for a host (the test caught it). Will need to dig into it a bit more to see how it can be handled.

**Comment by mattbrictson** [comment]
Unfortunately, any change we make here will likely break #1368. I am not 100% sure why, but for #1368 we made the decision to explicitly ignore username when declaring servers. In other words, the exact opposite of what is being requested for this issue.

I am guessing, but I believe the original reasoning behind #1368 comes from the discussion in #1301. Basically, some Capistrano users like the "additive" nature of `server` declarations, and it is more concise to merge server declaration by matching purely on hostname than by username@hostname.

Long story short, navigating a PR for this that makes everyone happy is going to extremely challenging.

If you do want to take a crack at a PR, please spend some time trying to understand the motivations behind #1301 and #1368 and see if those use cases can also be accommodated. You can also use #1368 as a reference for what files need to be changed.

**Comment by laszlof** [comment]
Right. I saw the code and it seemed what was being done was intentional. I'll see if I can work something out to accommodate both cases.

**Comment by laszlof** [comment]
ok. Read through the PR. I dont really agree with what was done there. The comment regarding keeping the stages "dry" doesnt make sense to me if we have to define our users in a task, or worse, in an ssh configuration.

That being said, I can understand that some people have different use cases, but it seems like this issue keeps cropping up where people want to do things the old way.

What are your thoughts on adding an option to provide support for the legacy method of explicitly matching against host, user, and port? maybe something like:

```
set :full_host_match, true
```

**Comment by mattbrictson** [comment]
Hmm. I would prefer not to add a setting.

Instead, what if we internally track a "key" for each server which is the string that was used to originally declare it. We always match on this key. (That means we could use a Hash, rather than looping through servers using `matches?`.)

For example:

```ruby
server "db01", :roles => %w(db) # key is "db01"
server "app@host", :roles => %w(app) # key is "app@host"

# This merges a new property onto db01 since the keys match
server "db01", :primary => true

# This creates a new server, since the keys do not match
server "web@host", :roles => %w(app)
```

Another example:

```ruby
# The key in both cases is "server1", so the properties are merged
role :redis, %w{server1 server2 server3 server4}, redis: { port: 6379, slave: true}
server 'server1', redis: { slave: false }
```

This might make everyone happy. Thoughts?

**Comment by laszlof** [comment]
That was one of my initial thoughts,  but may break existing installations
that rely on the existing functionality. But if your ok with that,  I can
probably work out a patch over the next couple days to make it happen.

**Comment by mattbrictson** [comment]
Yes, it will definitely break a handful of tests. Let's allow the discussion to continue for a bit longer before jumping into code.

@leehambley @ain @townsen Any thoughts on this?

**Comment by laszlof** [comment]
Fair enough!

**Comment by mattbrictson** [comment]
BTW, here's my initial take on a possible implementation:

```diff
diff --git a/lib/capistrano/configuration/servers.rb b/lib/capistrano/configuration/servers.rb
index 36d14d4..aa00889 100644
--- a/lib/capistrano/configuration/servers.rb
+++ b/lib/capistrano/configuration/servers.rb
@@ -8,13 +8,8 @@ module Capistrano
       include Enumerable
 
       def add_host(host, properties={})
-        new_host = Server[host]
-        if (server = servers.find { |s| s.matches? new_host })
-          server.user = new_host.user if new_host.user
-          server.with(properties)
-        else
-          servers << new_host.with(properties)
-        end
+        server = server_definitions[host.to_s] ||= Server[host]
+        server.with(properties)
       end
 
       def add_role(role, hosts, options={})
@@ -57,7 +52,11 @@ module Capistrano
       private
 
       def servers
-        @servers ||= []
+        server_definitions.values
+      end
+
+      def server_definitions
+        @server_definitions ||= {}
       end
 
       def extract_options(array)
```

**Comment by laszlof** [comment]
Seems pretty good from just looking over it, havent had a chance to test yet.

**Comment by laszlof** [comment]
Any further progress on this?

**Comment by mattbrictson** [comment]
I think there's been enough time for discussion and we can move ahead with a PR. Would you mind opening a PR? You can use [my suggestion](https://github.com/capistrano/capistrano/issues/1427#issuecomment-253698654) as a starting point if you like.

**Comment by laszlof** [comment]
Matt,

Sure, I'll go ahead and open a PR when I have some free time. Thanks!

**Comment by mattbrictson** [comment]
@laszlof Are you still planning on submitting a PR for this? No pressure, just curious! If you don't have time, let us know so that someone else can take a shot at it. Thanks!

**Comment by vevisystems** [comment]
In the meantime, we would like to share yet another workaround we have found, it's like @leehambley alternative but instead of using a SSH config file, you can set ssh options next to your server definitions:

server 'server_role_1', user: 'user1', roles: [...], ssh_options: { host_name: 'myserver.com' }
server 'server_role_2', user: 'user2', roles: [...], ssh_options: { host_name: 'myserver.com' }

**Comment by laszlof** [comment]
@mattbrictson Unfortunately my priorities have changed onto another project for now and have not had a chance to work on this. We have a workaround in place for the time being.

That being said, I think its an important feature and can get to it "some time." But if you have some spare bandwidth or someone else wants to take a crack at it feel free!
